README.md
# Isic

The International Standard Industrial Classification (ISIC) is a United Nations system for classifying economic data.

The classification is based in four hierarchical levels: sections, divisions, groups and classes.

This gem allows to classify an entity based on its ISIC code.

## Installation

Add this line to your application's Gemfile:

    gem 'isic'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install isic

## Usage

Given an ISIC code, find its place in the hierarchy:

```ruby
> Isic::Entity.new("0891").classify
{
  :section => { :code => "B", :description => "Mining and quarrying" },
  :division => { :code => "08", :description => "Other mining and quarrying" },
  :group => { :code => "089", :description => "Mining and quarrying n.e.c." },
  :class => { :code => "0891", :description => "Mining of chemical and fertilizer minerals" }
}
```

Find all the sections:

```ruby
> Isic.sections
[
  {:code=>"A", :description=>"Agriculture, forestry and fishing"},
  {:code=>"B", :description=>"Mining and quarrying"},
  {:code=>"C", :description=>"Manufacturing"}
  {:code=>"D", :description=>"Electricity, gas, steam and air conditioning supply"},
  {:code=>"E", :description=>"Water supply; sewerage, waste management and remediation activities"},
  {:code=>"F", :description=>"Construction"},
  {:code=>"G", :description=>"Wholesale and retail trade; repair of motor vehicles and motorcycles"},
  {:code=>"H", :description=>"Transportation and storage"},
  {:code=>"I", :description=>"Accommodation and food service activities"},
  {:code=>"J", :description=>"Information and communication"},
  {:code=>"K", :description=>"Financial and insurance activities"},
  {:code=>"L", :description=>"Real estate activities"},
  {:code=>"M", :description=>"Professional, scientific and technical activities"},
  {:code=>"N", :description=>"Administrative and support service activities"},
  {:code=>"O", :description=>"Public administration and defence; compulsory social security"},
  {:code=>"P", :description=>"Education"}, {:code=>"Q", :description=>"Human health and social work activities"},
  {:code=>"R", :description=>"Arts, entertainment and recreation"},
  {:code=>"S", :description=>"Other service activities"},
  {:code=>"T", :description=>"Activities of households as employers; undifferentiated goods- and services-producing activities of households for own use"},
  {:code=>"U", :description=>"Activities of extraterritorial organizations and bodies"}
]
```

Find all the divisions of the "B" section:

```ruby
> Isic.divisions(section: "B")
[
  {:code=>"05", :description=>"Mining of coal and lignite"},
  {:code=>"06", :description=>"Extraction of crude petroleum and natural gas"},
  {:code=>"07", :description=>"Mining of metal ores"},
  {:code=>"08", :description=>"Other mining and quarrying"},
  {:code=>"09", :description=>"Mining support service activities"}
]
```

Find all the groups of the "08" division:

```ruby
> Isic.groups(division: "08")
[
  {:code=>"081", :description=>"Quarrying of stone, sand and clay"},
  {:code=>"089", :description=>"Mining and quarrying n.e.c."}
]
```

Find all the classes of the "089" group:

```ruby
> Isic.classes(group: "089")
[
  {:code=>"0891", :description=>"Mining of chemical and fertilizer minerals"},
  {:code=>"0892", :description=>"Extraction of peat"},
  {:code=>"0893", :description=>"Extraction of salt"},
  {:code=>"0899", :description=>"Other mining and quarrying n.e.c."}
]
```

## Translations

Descriptions are returned in English by default, but Spanish and French are supported. In any of the methods above you can specify the translation:

```ruby
> Isic::Entity.new("0891").classify(translation: :es)
{
  :class=>{:code=>"0891", :description=>"Extracción de minerales para la fabricación de abonos y productos químicos"},
  :group=>{:code=>"089", :description=>"Explotación de minas y canteras n.c.p."},
  :division=>{:code=>"08", :description=>"Explotación de otras minas y canteras"},
  :section=>{:code=>"B", :description=>"Explotación de minas y canteras"}
}
```

```ruby
> Isic.groups(division: "08", translation: :fr)
[
  {:code=>"081", :description=>"Extraction de pierres, de sables et d'argiles"},
  {:code=>"089", :description=>"Activités extractives, n.c.a."}
]
```
## How to Launch the Console

    $ rake console

## How to Run the Test Suite

    $ rspec

## Changelog

[Changelog](https://github.com/javiervidal/isic/blob/master/CHANGELOG.md)

## Contributing

1. Fork it ( http://github.com/javiervidal/isic/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Code Status

[![Gem Version](http://img.shields.io/badge/gem-v1.0.4-blue.svg)](https://rubygems.org/gems/mic)
[![Build Status](https://travis-ci.org/javiervidal/isic.svg?branch=master)](https://travis-ci.org/javiervidal/isic)
[![Coverage Status](https://coveralls.io/repos/javiervidal/isic/badge.png?branch=master)](https://coveralls.io/r/javiervidal/isic?branch=master)
[![Code Climate](https://codeclimate.com/github/javiervidal/isic/badges/gpa.svg)](https://codeclimate.com/github/javiervidal/isic)

## Copyright

Copyright (c) 2014 – ∞ Javier Vidal

## License

This gem is released under the [MIT License](http://opensource.org/licenses/MIT).

CHANGELOG.md
# Change Log

## [1.0.5] - 2015-05-21

### Added
- This CHANGELOG.md file

### Changed
- Improve README.md
- Change .rvmrc for .ruby-version and .ruby-gemset in .gitignore

### Fixed
- [Fix files path ](https://github.com/javiervidal/isic/pull/1)

spec/isic_spec.rb
require 'spec_helper'

describe Isic do

  describe 'Isic methods' do

    context 'when english' do

      let(:sections) { Isic.sections }
      let(:divisions) { Isic.divisions(section: "B") }
      let(:groups) { Isic.groups(division: "08") }
      let(:classes) { Isic.classes(group: "089") }

      it 'returns all the sections in english' do
        expect(sections).to eq([{:code => "A", :description => "Agriculture, forestry and fishing"}, {:code => "B", :description => "Mining and quarrying"}, {:code => "C", :description => "Manufacturing"}, {:code => "D", :description => "Electricity, gas, steam and air conditioning supply"}, {:code => "E", :description => "Water supply; sewerage, waste management and remediation activities"}, {:code => "F", :description => "Construction"}, {:code => "G", :description => "Wholesale and retail trade; repair of motor vehicles and motorcycles"}, {:code => "H", :description => "Transportation and storage"}, {:code => "I", :description => "Accommodation and food service activities"}, {:code => "J", :description => "Information and communication"}, {:code => "K", :description => "Financial and insurance activities"}, {:code => "L", :description => "Real estate activities"}, {:code => "M", :description => "Professional, scientific and technical activities"}, {:code => "N", :description => "Administrative and support service activities"}, {:code => "O", :description => "Public administration and defence; compulsory social security"}, {:code => "P", :description => "Education"}, {:code => "Q", :description => "Human health and social work activities"}, {:code => "R", :description => "Arts, entertainment and recreation"}, {:code => "S", :description => "Other service activities"}, {:code => "T", :description => "Activities of households as employers; undifferentiated goods- and services-producing activities of households for own use"}, {:code => "U", :description => "Activities of extraterritorial organizations and bodies"}])
      end

      it 'returns all the divisions of the B section in english' do
        expect(divisions).to eq([{:code => "05", :description => "Mining of coal and lignite"}, {:code => "06", :description => "Extraction of crude petroleum and natural gas"}, {:code => "07", :description => "Mining of metal ores"}, {:code => "08", :description => "Other mining and quarrying"}, {:code => "09", :description => "Mining support service activities"}])
      end

      it 'returns all the groups of the 08 division in english' do
        expect(groups).to eq([{:code => "081", :description => "Quarrying of stone, sand and clay"}, {:code => "089", :description => "Mining and quarrying n.e.c."}])
      end

      it 'returns all the classes of the 089 group in english' do
        expect(classes).to eq([{:code => "0891", :description => "Mining of chemical and fertilizer minerals"}, {:code => "0892", :description => "Extraction of peat"}, {:code => "0893", :description => "Extraction of salt"}, {:code => "0899", :description => "Other mining and quarrying n.e.c."}])
      end

    end

    context 'when no specifying the parent entity' do

      let(:divisions) { Isic.divisions }
      let(:groups) { Isic.groups }
      let(:classes) { Isic.classes }

      it 'returns no divisions' do
        expect(divisions).to match_array([])
      end

      it 'returns no groups' do
        expect(groups).to match_array([])
      end

      it 'returns no classes' do
        expect(classes).to match_array([])
      end

    end

    context 'when spanish' do

      let(:sections) { Isic.sections(translation: :es) }
      let(:divisions) { Isic.divisions(section: "B", translation: :es) }
      let(:groups) { Isic.groups(division: "08", translation: :es) }
      let(:classes) { Isic.classes(group: "089", translation: :es) }

      it 'returns all the sections in english' do
        expect(sections).to eq([{:code => "A", :description => "Agricultura, ganadería, silvicultura y pesca"}, {:code => "B", :description => "Explotación de minas y canteras"}, {:code => "C", :description => "Industrias manufactureras"}, {:code => "D", :description => "Suministro de electricidad, gas, vapor y aire acondicionado"}, {:code => "E", :description => "Suministro de agua; evacuación de aguas residuales, gestión de desechos y descontaminación"}, {:code => "F", :description => "Construcción"}, {:code => "G", :description => "Comercio al por mayor y al por menor; reparación de vehículos automotores y motocicletas"}, {:code => "H", :description => "Transporte y almacenamiento"}, {:code => "I", :description => "Actividades de alojamiento y de servicio de comidas"}, {:code => "J", :description => "Información y comunicaciones"}, {:code => "K", :description => "Actividades financieras y de seguros"}, {:code => "L", :description => "Actividades inmobiliarias"}, {:code => "M", :description => "Actividades profesionales, científicas y técnicas"}, {:code => "N", :description => "Actividades de servicios administrativos y de apoyo"}, {:code => "O", :description => "Administración pública y defensa; planes de seguridad social de afiliación obligatoria"}, {:code => "P", :description => "Enseñanza"}, {:code => "Q", :description => "Actividades de atención de la salud humana y de asistencia social"}, {:code => "R", :description => "Actividades artísticas, de entretenimiento y recreativas"}, {:code => "S", :description => "Otras actividades de servicios"}, {:code => "T", :description => "Actividades de los hogares como empleadores; actividades no diferenciadas de los hogares como productores de bienes y servicios para uso propio"}, {:code => "U", :description => "Actividades de organizaciones y órganos extraterritoriales"}])
      end

      it 'returns all the divisions of the B section in english' do
        expect(divisions).to eq([{:code => "05", :description => "Extracción de carbón de piedra y lignito"}, {:code => "06", :description => "Extracción de petróleo crudo y gas natural"}, {:code => "07", :description => "Extracción de minerales metalíferos"}, {:code => "08", :description => "Explotación de otras minas y canteras"}, {:code => "09", :description => "Actividades de servicios de apoyo para la explotación de minas y canteras"}])
      end

      it 'returns all the groups of the 08 division in english' do
        expect(groups).to eq([{:code => "081", :description => "Extracción de piedra, arena y arcilla"}, {:code => "089", :description => "Explotación de minas y canteras n.c.p."}])
      end

      it 'returns all the classes of the 089 group in english' do
        expect(classes).to eq([{:code => "0891", :description => "Extracción de minerales para la fabricación de abonos y productos químicos"}, {:code => "0892", :description => "Extracción de turba"}, {:code => "0893", :description => "Extracción de sal"}, {:code => "0899", :description => "Explotación de otras minas y canteras n.c.p."}])
      end

    end

    context 'when french' do

      let(:sections) { Isic.sections(translation: :fr) }
      let(:divisions) { Isic.divisions(section: "B", translation: :fr) }
      let(:groups) { Isic.groups(division: "08", translation: :fr) }
      let(:classes) { Isic.classes(group: "089", translation: :fr) }

      it 'returns all the sections in english' do
        expect(sections).to eq([{:code => "A", :description => "Agriculture, sylviculture et pêche"}, {:code => "B", :description => "Activités extractives"}, {:code => "C", :description => "Activités de fabrication"}, {:code => "D", :description => "Production et distribution d'électricité, de gaz, de vapeur et climatisation"}, {:code => "E", :description => "Distribution d'eau; réseau d'assainissement; gestion des déchets et activités de remise en état"}, {:code => "F", :description => "Construction"}, {:code => "G", :description => "Commerce de gros et de détail, réparations de véhicules automobiles et de motocycles"}, {:code => "H", :description => "Transport et entreposage"}, {:code => "I", :description => "Activités d'hébergement et de restauration"}, {:code => "J", :description => "Information et communication"}, {:code => "K", :description => "Activités financières et d'assurances"}, {:code => "L", :description => "Activités immobilières"}, {:code => "M", :description => "Activités professionnelles, scientifiques et techniques"}, {:code => "N", :description => "Activités de services administratifs et d'appui"}, {:code => "O", :description => "Administration publique et défense; sécurité sociale obligatoire"}, {:code => "P", :description => "Éducation"}, {:code => "Q", :description => "Santé et et activités d'action sociale"}, {:code => "R", :description => "Arts, spectacles et loisirs"}, {:code => "S", :description => "Autres activités de services"}, {:code => "T", :description => "Activités des ménages privés employant du personnel domestique; activités non différenciées de production de biens et de services des ménages privés pour usage propre"}, {:code => "U", :description => "Activités des organisations et organismes extra-territoriaux"}])
      end

      it 'returns all the divisions of the B section in english' do
        expect(divisions).to eq([{:code => "05", :description => "Extraction de charbon et de lignite"}, {:code => "06", :description => "Extraction de pétrole brut et de gaz naturel"}, {:code => "07", :description => "Extraction de minerais métalliques"}, {:code => "08", :description => "Autres activités extractives"}, {:code => "09", :description => "Activités annexes de l'extraction"}])
      end

      it 'returns all the groups of the 08 division in english' do
        expect(groups).to eq([{:code => "081", :description => "Extraction de pierres, de sables et d'argiles"}, {:code => "089", :description => "Activités extractives, n.c.a."}])
      end

      it 'returns all the classes of the 089 group in english' do
        expect(classes).to eq([{:code => "0891", :description => "Extraction de minerais pour l'industrie chimique et d'engrais naturels"}, {:code => "0892", :description => "Extraction de tourbe"}, {:code => "0893", :description => "Extraction de sel"}, {:code => "0899", :description => "Autres activités extractives, n.c.a."}])
      end

    end

  end

end

spec/entity_spec.rb
require 'spec_helper'

describe Isic::Entity do

  describe '#classify' do

    context 'in english' do

      let(:class_classification) { Isic::Entity.new("0891").classify }
      let(:group_classification) { Isic::Entity.new("089").classify }
      let(:division_classification) { Isic::Entity.new("08").classify }
      let(:section_classification) { Isic::Entity.new("B").classify }

      it "class is classified with section, division, group and class" do
        expect(class_classification).to eq({
                                             :class => {:code => "0891", :description => "Mining of chemical and fertilizer minerals"},
                                             :group => {:code => "089", :description => "Mining and quarrying n.e.c."},
                                             :division => {:code => "08", :description => "Other mining and quarrying"},
                                             :section => {:code => "B", :description => "Mining and quarrying"}
                                           })
      end

      it "group is classified with section, division and group" do
        expect(group_classification).to eq({
                                             :group => {:code => "089", :description => "Mining and quarrying n.e.c."},
                                             :division => {:code => "08", :description => "Other mining and quarrying"},
                                             :section => {:code => "B", :description => "Mining and quarrying"}
                                           })
      end

      it "division is classified with section and division" do
        expect(division_classification).to eq({
                                                :division => {:code => "08", :description => "Other mining and quarrying"},
                                                :section => {:code => "B", :description => "Mining and quarrying"}
                                              })
      end

      it "section is classified with section" do
        expect(section_classification).to eq({
                                               :section => {:code => "B", :description => "Mining and quarrying"}
                                             })
      end

    end

    context 'in spanish' do

      let(:class_classification) { Isic::Entity.new("0891").classify(translation: :es) }

      it "classifies with section, division, group and class" do
        expect(class_classification).to eq({
                                             :class => {:code => "0891", :description => "Extracción de minerales para la fabricación de abonos y productos químicos"},
                                             :group => {:code => "089", :description => "Explotación de minas y canteras n.c.p."},
                                             :division => {:code => "08", :description => "Explotación de otras minas y canteras"},
                                             :section => {:code => "B", :description => "Explotación de minas y canteras"}
                                           })
      end

    end

    context 'in french' do

      let(:class_classification) { Isic::Entity.new("0891").classify(translation: :fr) }

      it "classifies with section, division, group and class" do
        expect(class_classification).to eq({
                                             :class => {:code => "0891", :description => "Extraction de minerais pour l'industrie chimique et d'engrais naturels"},
                                             :group => {:code => "089", :description => "Activités extractives, n.c.a."},
                                             :division => {:code => "08", :description => "Autres activités extractives"},
                                             :section => {:code => "B", :description => "Activités extractives"}
                                           })
      end

    end

  end

end

spec/spec_helper.rb
require 'simplecov'
require 'coveralls'

SimpleCov.formatters = [SimpleCov::Formatter::HTMLFormatter, Coveralls::SimpleCov::Formatter]

SimpleCov.start

require 'isic'

RSpec.configure do |config|

  config.expect_with :rspec do |expectations|
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end

end

lib/isic/entity.rb
class Isic
  class Entity

    def initialize(code)
      @code = code
    end

    def classify(options = {})
      translation = options[:translation] || :en
      all_codes.inject({}) do |hash, (key, value)|
        hash[key] = Isic::Search.new(value, translation: translation).first
        hash
      end
    end

    private

    def all_codes
      case @code
        when /\d{4}/
          {class: @code, group: @code[0..2], division: @code[0..1], section: section(@code[0..1])}
        when /\d{3}/
          {group: @code, division: @code[0..1], section: section(@code[0..1])}
        when /\d{2}/
          {division: @code, section: section(@code)}
        when /[A-Z]/
          {section: @code}
      end
    end

    def section(division)
      Isic::DIVISIONS.select { |k,v| v.include?(division) }.keys.first
    end

  end
end

lib/isic/search.rb
class Isic
  class Search

    FILES = {
      en: File.join(File.dirname(File.expand_path(__FILE__)), '../../files/ISIC_Rev_4_english_structure.txt'),
      es: File.join(File.dirname(File.expand_path(__FILE__)), '../../files/ISIC_Rev_4_spanish_structure.txt'),
      fr: File.join(File.dirname(File.expand_path(__FILE__)), '../../files/ISIC_Rev_4_french_structure.txt')
    }

    ENCODINGS = {
      en: 'utf-8',
      es: 'iso-8859-15:utf-8',
      fr: 'iso-8859-15:utf-8'
    }

    def initialize(regexp, translation: :en)
      @regexp = regexp
      @file = FILES[translation]
      @encoding = ENCODINGS[translation]
    end

    def all
      entities = []
      File.open(@file, encoding: @encoding).each do |line|
        md = /"(#{@regexp})","(.+)"/.match(line)
        entities << {code: md[1], description: md[2]} if md
      end
      entities
    end

    def first
      all.first
    end

  end
end

lib/isic/isic.rb
class Isic

  DIVISIONS = {}
  DIVISIONS['A'] = %w( 01 02 03 )
  DIVISIONS['B'] = %w( 05 06 07 08 09 )
  DIVISIONS['C'] = %w( 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 )
  DIVISIONS['D'] = %w( 35 )
  DIVISIONS['E'] = %w( 36 37 38 39 )
  DIVISIONS['F'] = %w( 41 42 43 )
  DIVISIONS['G'] = %w( 45 46 47 )
  DIVISIONS['H'] = %w( 49 50 51 52 53 )
  DIVISIONS['I'] = %w( 55 56 )
  DIVISIONS['J'] = %w( 58 59 60 61 62 63 )
  DIVISIONS['K'] = %w( 64 65 66 )
  DIVISIONS['L'] = %w( 68 )
  DIVISIONS['M'] = %w( 69 70 71 72 73 74 75 )
  DIVISIONS['N'] = %w( 77 78 79 80 81 82 )
  DIVISIONS['O'] = %w( 84 )
  DIVISIONS['P'] = %w( 85 )
  DIVISIONS['Q'] = %w( 86 87 88 )
  DIVISIONS['R'] = %w( 90 91 92 93 )
  DIVISIONS['S'] = %w( 94 95 96 )
  DIVISIONS['T'] = %w( 97 98 )
  DIVISIONS['U'] = %w( 99 )

  class << self

    def sections(options = {})
      translation = options[:translation] || :en
      Search.new('[A-Z]', translation: translation).all
    end

    def divisions(options = {})
      translation = options[:translation] || :en
      if options[:section] && DIVISIONS.keys.include?(options[:section])
        code = DIVISIONS[options[:section]].join('|')
        Search.new(code, translation: translation).all
      else
        []
      end
    end

    def groups(options = {})
      parse(:division, 2, options)
    end

    def classes(options = {})
      parse(:group, 3, options)
    end

    def parse(symbol, digits, options = {})
      translation = options[:translation] || :en
      if options[symbol] && Regexp.new("\\d{#{digits}}").match(options[symbol])
        code = "#{options[symbol]}\\d"
        Search.new(code, translation: translation).all
      else
        []
      end
    end

  end

  private_class_method :parse

end